app/models/ems_refresh/save_inventory_helper.rb

```diff
@@ -113,12 +113,28 @@ def save_child_inventory(obj, hashes, child_keys, *args)
 
   def store_ids_for_new_records(records, hashes, keys)
     keys = Array(keys)
-    hashes.each do |h|
-      r = records.detect { |r| keys.all? { |k| r.send(k) == r.class.type_for_attribute(k.to_s).cast(h[k]) } }
-      h[:id] = r.id
+    # Lets first index the hashes based on keys, so we can do O(1) lookups
+    record_index = {}
+    records.each do |record|
```

Comment:
Could `records.select(:id, *keys)` help further?
Also, we don't need record_index to store the records themselves, can directly store the id.